### ISSUE TITLE
Expect.toThrow for async functions fails when an error is thrown

### ISSUE THREAD

**Issue by winksaville** [alsatian-test/alsatian#353]
I've created this [test code](https://github.com/winksaville/test-Alsatian-Expect-toThrow-async). Here is a snippet of the failing test:
```
    // Asynchronous throw
    public async asyncThrowError(): Promise<void> {
        return new Promise<void>(() => {
            setTimeout(Promise.reject(new Error("Timeout then reject")), 100);
        });
    }

    @AsyncTest("test alsatian async toThrow")
    public async testAsyncToThrow() {
        debug("testToThrow:+");
        Expect(async () => await this.asyncThrowError()).toThrow();
        debug("testToThrow:-");
    }
```
And here is the output when using --tap:
```
testAsyncToThrow:+
not ok 2 test alsatian async toThrow
(node:15093) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: Timeout then reject
(node:15093) DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
(node:15093) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 6): TypeError: "callback" argument must be a function
 ---
   message: "Expected an error to be thrown but no errors were thown."
   severity: fail
   data:
     got: "error was not thrown."
     expect: "error to be thrown."
 ...
```

**Comment by winksaville** [comment]
As a work around I'm doing the following:
```
    @AsyncTest("test alsatian async")
    public async testAsync() {
        debug("testAsync:+");
        try {
            await this.asyncThrowError();
            Expect("No error thrown, fail").not.toBeTruthy();
        } catch(err) {
            Expect(err).toBeTruthy();
        }
        debug("testAsync:-");
    }
```

**Comment by jamesrichford** [comment]
@winksaville - very interesting use case indeed!!

I'll check out the PR later to find out more :)

**Comment by winksaville** [comment]
@jamesrichford, What do you think?

**Comment by jamesrichford** [comment]
Hey @winksaville,

I think that it's a cool idea and we'd definitely want to have something along those lines - just need to experiment a bit to see what options we have I think :) I'm hoping to have a bit of time this weekend to experiment. From what I've seen of yours you'd need to await the Expect is that correct?

**Comment by winksaville** [comment]
My solution was to duplicate toThrow and call it toThrowAsync and then adding  an `await` in the try/catch, no other changes:
```
   public async toThrowAsync() {

      if (this._actualValue instanceof Function === false) {
         throw new TypeError("toThrow requires value passed in to Expect to be a function.");
      }

      let errorThrown: Error;

      try {
         await this._actualValue();  ///<< Added await here
      }
      catch (error) {
         errorThrown = error;
      }

      if (errorThrown === undefined === this.shouldMatch) {
         throw new ErrorMatchError(errorThrown, this.shouldMatch);
      }
   }
```
What I wanted to do was this, but it didn't work:
```
Expect(async () => await xyz()).toThrow();
```
or
```
Expect(async () => await xyz()).not.toThrow();
```
or
```
Expect(async () => await xyz().toBeTruthy()
```
...

**Comment by jamesrichford** [comment]
@winksaville sure :) I can see that.

What I rather meant was would this?

```typescript
await Expect(async.function).toThrowAsync();
```

OR

```typescript
Expect(async.function).toThrowAsync();
```

**Comment by winksaville** [comment]
@jamesrichford, whatever works, I think a requirement is that `toThrowAsync` needs to `await` inside the `try` if `this._actualValue() returns a Promise. Otherwise the `catch` never occurs before if statement after the try/catch:
```
      if (errorThrown === undefined === this.shouldMatch) {
```
And errorThrown is always undefined.

Maybe a solution could be (note: probably BS as I'm a typescript neophyte):
```
    consturctor(actualValue: () : Promise<any>) {
        this._actualValueAsync = actualValue;
    }
    constructor(actualValue: any)
        this._actualValue = actualValue;
    }
...
   public async toThrow() {

      if (this._actualValue instanceof Function === false) {
         throw new TypeError("toThrow requires value passed in to Expect to be a function.");
      }

      let errorThrown: Error;

      try {
         if (this._actualValue) {
             this._actualValue();
         } else {
             await this._actualValueAsync();
          }
      }
      catch (error) {
         errorThrown = error;
      }

      if (errorThrown === undefined === this.shouldMatch) {
         throw new ErrorMatchError(errorThrown, this.shouldMatch);
      }
   }
```

**Comment by jamesrichford** [comment]
@winksaville probably thinking about it (need to experiment to prove this) but there should be no way of having this behaviour inside a normal test as the runtime wouldn't know when the function has completed so this would always have to be within an async test so perhaps going with the toThrowAsync would be better as it's more clear that you need to run it within an async test.

**Comment by winksaville** [comment]
SG, but I'm guessing other matchers are going to need `xxxAsync` too as the returned value from any `async` function inside an `Expect` will be returning a Promise if the `await` isn't performed. Remember this is a neophyte talking crap.

**Comment by jamesrichford** [comment]
@winksaville I've had a deeper look now and looks like this would be the proposed usage

```typescript
import { Expect, SpyOn, AsyncTest, Timeout, TestFixture } from "alsatian";

class ObjectUnderTest {
    public async function() {
        return new Promise((resolve, reject) => {
            setTimeout(() => {
                reject(new Error("example async error"));
            }, 2000);
        })
    }
}

@TestFixture("async error example fixture")
export class AsyncErrorExample {

   @Timeout(4000)
   @AsyncTest("async error example")
   public async asyncErrorExample() {
       const a = new ObjectUnderTest();

       SpyOn(a, "function");

       await Expect(a.function).toThrowAsync();
   }
}
```

The one thing that may be problematic slightly is that the `await` keyword is essential on this expectation. We could support with documentation but as things stand if the user does not use the await keyword the test will report a false positive.

Perhaps as part of alsatian 2.0 we can consider avoiding these false positives but would you say the above is acceptable for now? If so we can crack on with a fuller version which should include `toThrowErrorAsync` and relevant tests and documentation :)

**Comment by winksaville** [comment]
I assume using SpyOn isn't necessary?

I'm curious how its going to work I think the await is still needed inside `toThrowAsync` as in my hack, so I'm wondering why its needed outside?

Assuming the `await` is needed its OK with me.

**Comment by jamesrichford** [comment]
The outer `await` is necessary because without it the functionality would return before the asynchronous has resolved. The inner `await` will also be necessary but the test itself will need to know it's waiting on that one hence the outer `await`.

You could also use the `return` keyword which would have the same affect but `await` is clearer

**Comment by winksaville** [comment]
Updated my [Hack](https://github.com/alsatian-test/alsatian/pull/354) to have toThrowAsync return a
Promise<void> and my [test-Alsatian-Expect-toThrow-async](https://github.com/winksaville/test-Alsatian-Expect-toThrow-async) tested it and the output now looks better!

**Comment by jamesrichford** [comment]
Nice one @winksaville.

Is this PR serving as an example or are we aiming to have this merged into master?

**Comment by winksaville** [comment]
@jamesrichford, up to you, if its on the right track we can continue or you can use it as a proof of concept and do it "right". You're welcome to take it from here or I can spend a little more time on it. Obviously there are no docs so that needs to be done.

**Comment by winksaville** [comment]
What would you like me to do?

**Comment by jamesrichford** [comment]
Sorry @winksaville I've been a bit busy of late so not able to add a nice bit of detail. I'll try and build some up for you now

**Comment by jamesrichford** [comment]
TO DO:

- [ ] implement `toThrowAsync()` in the same manner as `toThrow()` but awaits the given function
- [ ] implement `toThrowErrorAsync(type, message)` in the same manner as `toThrowError()` but awaits the given function
- [ ] pass all CI checks
- [ ] add integration tests
- [ ] consider refactors with existing synchronous functions

**Comment by jamesrichford** [comment]
@winksaville I've made a check list above of the high level stuff that needs to be fine for this. If you need any help out more detail please let me know :)

Thanks

**Comment by winksaville** [comment]
Got it

**Comment by winksaville** [comment]
I've updated PR #354 please review. I did not add integration tests as I couldn't run them as install-self appears to want to install them globall. Is there a way to run them without installing globally?

**Comment by jamesrichford** [comment]
Hi @winksaville. What do you mean by installs them globally? All install-self does is adds a reference back to the local project in the local `node_modules` folder.

**Comment by winksaville** [comment]
Here is the output:
```
 core/stringification                 |      100 |      100 |      100 |
   100 |                |
  argument-stringifier.ts             |      100 |      100 |      100 |
   100 |                |
  index.ts                            |      100 |      100 |      100 |
   100 |                |
--------------------------------------|----------|----------|----------|----------|----------------|

> alsatian@1.3.0 integration-tests /home/wink/prgs/alsatian.myfork
> npm run install-self && npm run cli-integration-tests && npm run
node-integration-tests && npm run gulp-integration-tests&& npm run
babel-integration-tests && npm run expect-extension-integration-tests


> alsatian@1.3.0 install-self /home/wink/prgs/alsatian.myfork
> npm link && install-self

npm ERR! path /usr/lib/node_modules/alsatian
npm ERR! code EACCES
npm ERR! errno -13
npm ERR! syscall unlink
npm ERR! Error: EACCES: permission denied, unlink
'/usr/lib/node_modules/alsatian'
npm ERR!  { Error: EACCES: permission denied, unlink
'/usr/lib/node_modules/alsatian'
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'unlink',
npm ERR!   path: '/usr/lib/node_modules/alsatian' }
npm ERR!
npm ERR! Please try running this command again as root/Administrator.
```

And I was mistaken, it's the `npm link` that's generating the error, here
is me executing it manually from the command line:
```
$ npm link
npm ERR! path /usr/lib/node_modules/alsatian
npm ERR! code EACCES
npm ERR! errno -13
npm ERR! syscall unlink
npm ERR! Error: EACCES: permission denied, unlink
'/usr/lib/node_modules/alsatian'
npm ERR!  { Error: EACCES: permission denied, unlink
'/usr/lib/node_modules/alsatian'
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'unlink',
npm ERR!   path: '/usr/lib/node_modules/alsatian' }
npm ERR!
npm ERR! Please try running this command again as root/Administrator.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/wink/.npm/_logs/2017-04-26T13_32_07_183Z-debug.log
```

**Comment by winksaville** [comment]
Note the [npmjs docs](https://docs.npmjs.com/cli/link) say `npm link` adds a link to the `global folder`:
```
First, npm link in a package folder will create a symlink in the global folder {prefix}/lib/node_modules/<package> that links to the package where the npm link command was executed.
```
Which needs `root permissions` on my machine.

**Comment by winksaville** [comment]
If I modify package.json to remove all references to `alsatian` to be `node ./cli/alsatian-cli.js` and remove invoking `install-self`:
```
$ git diff package.json
diff --git a/package.json b/package.json
index 902ba0c..3088537 100644
--- a/package.json
+++ b/package.json
@@ -40,16 +40,16 @@
     "check-coverage": "npm run build & nyc --reporter=lcov --reporter=html npm run unit-tests && npm run output-test-coverage",
     "output-test-coverage": "nyc report",
     "install-self": "npm link && install-self",
-    "integration-tests": "npm run install-self && npm run cli-integration-tests && npm run node-integration-tests && npm run gulp-integration-tests&& npm run babel-integration-tests && npm run expect-extension-integration-tests",
-    "expect-extension-integration-tests": "alsatian \"./test/integration-tests/extending-expect/**/*.spec.js\"",
-    "cli-integration-tests": "alsatian \"./test/integration-tests/cli/runner\"",
+    "integration-tests": "npm run cli-integration-tests && npm run node-integration-tests && npm run gulp-integration-tests&& npm run babel-integration-tests && npm run expect-extension-integration-tests",
+    "expect-extension-integration-tests": "node ./cli/alsatian-cli.js \"./test/integration-tests/extending-expect/**/*.spec.js\"",
+    "cli-integration-tests": "node ./cli/alsatian-cli.js \"./test/integration-tests/cli/runner\"",
     "node-integration-tests": "tsc -p \"./test/integration-tests\" && node \"./test/integration-tests/node/runner\"",
-    "gulp-integration-tests": "alsatian \"./test/integration-tests/gulp/runner\"",
+    "gulp-integration-tests": "node ./cli/alsatian-cli.js \"./test/integration-tests/gulp/runner\"",
     "debug-unit-tests": "node-debug ./cli/alsatian-cli.js \"./test/unit-tests/**/*.spec.js\"",
     "build-es6-tests": "babel test/integration-tests/javascript/test-sets -d test/integration-tests/javascript/test-sets",
-    "babel-integration-tests": "npm run build-es6-tests && alsatian \"./test/integration-tests/javascript/babel/babel-compatibility.spec\"",
-    "typescript-integration-tests": "alsatian \"./test/integration-tests/typescript/compile.spec\"",
-    "selenium-integration-tests": "alsatian \"./test/integration-tests/selenium/**/*.spec.js\""
+    "babel-integration-tests": "npm run build-es6-tests && node ./cli/alsatian-cli.js \"./test/integration-tests/javascript/babel/babel-compatibility.spec\"",
+    "typescript-integration-tests": "node ./cli/alsatian-cli.js \"./test/integration-tests/typescript/compile.spec\"",
+    "selenium-integration-tests": "node ./cli/alsatian-cli.js \"./test/integration-tests/selenium/**/*.spec.js\""
   },
   "repository": {
     "type": "git",
```
The integration tests run:
```
 core/stringification                 |      100 |      100 |      100 |      100 |                |
  argument-stringifier.ts             |      100 |      100 |      100 |      100 |                |
  index.ts                            |      100 |      100 |      100 |      100 |                |
--------------------------------------|----------|----------|----------|----------|----------------|

> alsatian@1.3.0 integration-tests /home/wink/prgs/alsatian.myfork
> npm run cli-integration-tests && npm run node-integration-tests && npm run gulp-integration-tests&& npm run babel-integration-tests && npm run expect-extension-integration-tests


> alsatian@1.3.0 cli-integration-tests /home/wink/prgs/alsatian.myfork
> node ./cli/alsatian-cli.js "./test/integration-tests/cli/runner"
CliIntegrationTests
syntaxTests ( "case-arguments" )
|====================|

Pass: 5/5
Fail: 0/5
Ignore: 0/5

> alsatian@1.3.0 node-integration-tests /home/wink/prgs/alsatian.myfork
> tsc -p "./test/integration-tests" && node "./test/integration-tests/node/runner"


> alsatian@1.3.0 gulp-integration-tests /home/wink/prgs/alsatian.myfork
> node ./cli/alsatian-cli.js "./test/integration-tests/gulp/runner"
GulpIntegrationTests
syntaxTests ( "teardown" )
|====================|

Pass: 4/4
Fail: 0/4
Ignore: 0/4

> alsatian@1.3.0 babel-integration-tests /home/wink/prgs/alsatian.myfork
> npm run build-es6-tests && node ./cli/alsatian-cli.js "./test/integration-tests/javascript/babel/babel-compatibility.spec"


> alsatian@1.3.0 build-es6-tests /home/wink/prgs/alsatian.myfork
> babel test/integration-tests/javascript/test-sets -d test/integration-tests/javascript/test-sets

test/integration-tests/javascript/test-sets/expectations/to-be.spec.es6 -> test/integration-tests/javascript/test-sets/expectations/to-be.spec.js
test/integration-tests/javascript/test-sets/expectations/to-be.spec.js -> test/integration-tests/javascript/test-sets/expectations/to-be.spec.js
test/integration-tests/javascript/test-sets/test-syntax/async-test.spec.es6 -> test/integration-tests/javascript/test-sets/test-syntax/async-test.spec.js
test/integration-tests/javascript/test-sets/test-syntax/async-test.spec.js -> test/integration-tests/javascript/test-sets/test-syntax/async-test.spec.js
test/integration-tests/javascript/test-sets/test-syntax/case-arguments.spec.es6 -> test/integration-tests/javascript/test-sets/test-syntax/case-arguments.spec.js
test/integration-tests/javascript/test-sets/test-syntax/case-arguments.spec.js -> test/integration-tests/javascript/test-sets/test-syntax/case-arguments.spec.js
test/integration-tests/javascript/test-sets/test-syntax/setup.spec.es6 -> test/integration-tests/javascript/test-sets/test-syntax/setup.spec.js
test/integration-tests/javascript/test-sets/test-syntax/setup.spec.js -> test/integration-tests/javascript/test-sets/test-syntax/setup.spec.js
test/integration-tests/javascript/test-sets/test-syntax/teardown-fixture.spec.es6 -> test/integration-tests/javascript/test-sets/test-syntax/teardown-fixture.spec.js
test/integration-tests/javascript/test-sets/test-syntax/teardown-fixture.spec.js -> test/integration-tests/javascript/test-sets/test-syntax/teardown-fixture.spec.js
test/integration-tests/javascript/test-sets/test-syntax/teardown-tests.spec.es6 -> test/integration-tests/javascript/test-sets/test-syntax/teardown-tests.spec.js
test/integration-tests/javascript/test-sets/test-syntax/teardown-tests.spec.js -> test/integratiBabelIntegrationTests
asyncTest
|====================|

Pass: 2/2
Fail: 0/2
Ignore: 0/2

> alsatian@1.3.0 expect-extension-integration-tests /home/wink/prgs/alsatian.myfork
> node ./cli/alsatian-cli.js "./test/integration-tests/extending-expect/**/*.spec.js"
ExtendingExpectTests
extension retains existing functions
|====================|

Pass: 3/3
Fail: 0/3
Ignore: 0/3
```

**Comment by winksaville** [comment]
Jeez, I just looked at one of the test/integration-tests/cli/runner.ts and I see and others reference the global `alsatian` code:
```
$ find . -type f -name '*.ts' | xargs grep 'exec\>.*alsatian'
./test/integration-tests/cli/runner.ts:                    .exec(`alsatian ` +
./test/integration-tests/cli/runner.ts:                    .exec(`alsatian ` +
./test/integration-tests/javascript/babel/babel-compatibility.spec.ts:                .exec("alsatian ./test/integration-tests/javascript/test-sets/expectations/to-be.spec.js --tap");
./test/integration-tests/javascript/babel/babel-compatibility.spec.ts:                    .exec("alsatian \"./test/integration-tests/javascript/test-sets/test-syntax/**/*.spec.js\" --tap");
```
What the heck, I'll change them too and see what happens :)

**Comment by winksaville** [comment]
Not sure you want it but I've pushed a new change that modifies package.json and integration tests to not reference `alsatian` and instead execute it via `node ./cli/alsatian-cli.js` things seem to be working. I also note that there are no tests of `toThrow` in the integration tests, well I don't see any greping for `toThrow`:
```
$ find . -type f -name '*' | xargs grep 'toThrow'
$
```
So are there some specific integration tests you'd like me to add?

**Comment by Jameskmonger** [comment]
I much prefer this @jamesrichford - I don't think there's a need to have it self installed if we can just call the CLI with `node`.

**Comment by jamesrichford** [comment]
Sorry put this in the PR by mistake :)

@Jameskmonger @winksaville the reason for this is it is an integration test and checking whether the code works together including the package.json referring to the main and bin files correctly. I think the problem here is maybe the link is redundant but the install-self should still be happening as this should be referencing the local version itself. Can this be checked please?

**Comment by Jameskmonger** [comment]
Maybe we should have a separate package, `alsatian-integration` which runs the integration tests. That way `alsatian` can be a proper dependency and follow the normal npm install process rather than relying on `install-self` (which doesn't appear to do a proper npm install but rather just fake the npm install by pointing at the local code through a `package.json`).

**Comment by jamesrichford** [comment]
@Jameskmonger that'd mean managing 2 separate repos for 1 project to ensure integration tests and alsatian itself are in line. Maybe in the future but it requires lots of additional work where as this currently emulates the process quite well. The other problem is what do we do when we want to run integration tests before publishing, which is key for ensuring we don't deliver a broken version. We can never npm install a version that doesn't exist (particularly if it's coming in from a PR) so separating that out would require a bit of thought.

I think this gives the right balance in level of emulation in that how npm will run the script is trace package.json files which is what's happening but ensuring that we're pointing to the new version and not the latest distributed version.

**Comment by Jameskmonger** [comment]
What we could do @jamesrichford is make a new `package.json` in `/integration-tests/` which installs alsatian as a local reference. That way they'll always be in sync.

**Comment by winksaville** [comment]
I'd like to say that the solution of using `alsatian-cli.js` directly means there is only one version used and referenced only one way. My 2 cents.

**Comment by winksaville** [comment]
So with my latest changes the everything passed on Appveyor but there was one timeout failure on Travis:
```
Pass: 3/4
Fail: 1/4
Ignore: 0/4
FAIL: toBeExpectations
The test exceeded the given timeout of 5000ms.
Expected: undefined
  Actual: undefined
```

Suggestions on next steps?

**Comment by jamesrichford** [comment]
@winksaville referencing it that way is certainly the simplest and safest. However the integration tests should sill be referencing the same version of the same file via that mechanism. The point is to be referencing it via a `package.json` to better simulate what a user will do. So even though we're not referencing the file directly we are indirectly via the install-self functionality (which adds a reference to the projects `package.json`. Probably could do with some improvement but out of scope for now :)

With regards to your issue I would say this is probably due to the speed issues we're seeing on certain configurations of travis. I'll have a look into it and let you know. As far as your code is concerned from the sound of it I would say that it's nothing to do with what you've written :)

**Comment by winksaville** [comment]
I'll the addition of the new `package.json` to you guys and I'll revert my tweaks to fix intergration-tests. But I'll leave in my bump to the timeout which fixed my last problem.

**Comment by winksaville** [comment]
As you can see I added PR #370 which doesn't have the integration-test changes. Let me know what else you'd like me to do.
